src/device/device_default.hpp

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2017-2018, The Monero Project
+// 
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+// 
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+// 
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#pragma once
+
+#include "device.hpp"
+
+namespace hw {
+
+    namespace core {
+
+        void register_all();
+
+        class DeviceDefault : public hw::Device {
+        public:
+            DeviceDefault();
+            DeviceDefault(const DeviceDefault &device);
+            ~DeviceDefault();
+
+            DeviceDefault& operator=(const DeviceDefault &device);
+            explicit operator bool() const { return false; };
+
+             /* ======================================================================= */
+            /*                              SETUP/TEARDOWN                             */
+            /* ======================================================================= */
+
+
+            /* ======================================================================= */
+            /*                             WALLET & ADDRESS                            */
+            /* ======================================================================= */
+            //bool  get_public_address(cryptonote::account_public_address &pubkey);
+            #ifdef DEBUGLEDGER
+            //bool  get_secret_keys(crypto::secret_key &viewkey , crypto::secret_key &spendkey);
+            #endif
+            bool  generate_chacha_key(const cryptonote::account_keys &keys, crypto::chacha_key &key);
```

Comment:
Theres lots to add in here, but add in the `override` tag to all of thee `virtual` functions so the compiler errors out if the base class interface changes. It **will** do this if the base class functions are pure virtual (because a definition is required), but otherwise this keyword helps subtle issues with interface changes.

Reply:
override added everywhere in device_default and device_ledger